Tone/source/ExternalInput.js

```diff
@@ -0,0 +1,220 @@
+
+		if(Tone.ExternalInput.sourceList.length === 0){
+			Tone.ExternalInput._getSources();
+		} 
+
+		/**
+		 * The input source position in Tone.ExternalInput.sourceList. 
+		 * Set before ExternalInput.start().
+		 * @type {[type]}
+		 */
+		this.inputNum = options.inputNum;
+
+	};
+
+	Tone.extend(Tone.ExternalInput, Tone.Source);
+
+	/**
+	 * the default parameters
+	 * @type {Object}
+	 */
+	Tone.ExternalInput.defaults = {
+		"inputNum" : 0,
+		"onload" : Tone.noOp
+	};
+
+	/**
+	 * Start the stream
+	 * @private
+	 */
+	Tone.ExternalInput.prototype._start = function(){
+		if(!Tone.ExternalInput.sourceList.length && Tone.ExternalInput._hasGetSources && this.inputNum){
+			console.error("Default source selected, use ExternalInput.onload to allow source loading");
+		}
+		if(!Tone.ExternalInput._hasGetSources && this.inputNum > 0){
+			console.error("This browser does not support MediaStreamTrack sourceId");
+		}
+		if (this.inputNum < Tone.ExternalInput.sourceList.length && this.inputNum > -1) {
+			this._constraints = {
+				audio : {
+					optional : [{ sourceId: Tone.ExternalInput.sourceList[this.inputNum].id}]
+				}
+			};
+		} else {
+			console.error("inputNum was outside of sourceList bounds, default source selected");
+		}
+		navigator.getUserMedia(this._constraints, 
+			this._onStream.bind(this), this._onStreamError.bind(this));
+	};
+
+	/**
+	 * Stops the stream. Stream can be restarted with a different inputNum.
+	 * @private
+	 */
+	Tone.ExternalInput.prototype._stop = function(){
+		this._stream.stop();
+		return this;
+	};
+
+	/**
+	 * called when the stream is successfully setup
+	 * @param  {LocalMediaStream} stream
+	 * @private
+	 */
+	Tone.ExternalInput.prototype._onStream = function(stream){
+		this._stream = stream;
+		// Wrap a MediaStreamSourceNode around the live input stream.
+		this._mediaStream = this.context.createMediaStreamSource(stream);
+		this._mediaStream.connect(this.output);
+	};
+	/**
+	 * called on error
+	 * @param  {Error} e 
+	 * @private
+	 */
+	Tone.ExternalInput.prototype._onStreamError = function(e){
+		console.error(e);
+	};
+
+	/**
+	 * Clean up.
+	 * @return {Tone.Microphone} this
+	 */
+	Tone.ExternalInput.prototype.dispose = function(){
+		Tone.Source.prototype.dispose.call(this);
+		if (this._mediaStream){
+			this._mediaStream.disconnect();
+			this._mediaStream = null;
+		}
+		this._stream = null;
+		this._constraints = null;
+		return this;
+	};
+
+	///////////////////////////////////////////////////////////////////////////
+	// STATIC METHODS
+	///////////////////////////////////////////////////////////////////////////
+
+	/**
+	 * The array of available sources, different depending on whether connection is secure
+	 * @type {Array}
+	 */
+	Tone.ExternalInput.sourceList = [];
+
+	/**
+	 * indicates whether browser supports MediaStreamTrack.getSources (i.e. Chrome vs Firefox)
+	 * @type {Boolean}
+	 * @private
+	 */
+	Tone.ExternalInput._hasGetSources = true;
```

Comment:
You could check if it has the `getSources` method here. instead of once you call `_getSources` instead of setting it to true initially. 

```
Tone.ExternalInput._hasGetSources = Tone.prototype.isFunction(MediaStreamTrack.getSources)
```

also not really into the name `hasGetSources`, maybe `canGetSources`?